docs/tutorials/dcdi/sniffing/custom.py

```diff
@@ -0,0 +1,39 @@
+"""
+
+A custom add-on to our app which adds FrenchCustomer and
+French Greeter.
+
+"""
+from dataclasses import dataclass
+
+from wired import ServiceRegistry
+from .models import Customer, Greeter, Datastore
+from .utils import register_dataclass
+
+
+@dataclass
+class FrenchCustomer(Customer):
+    """ A custom kind of Customer """
+
+    pass
+
+
+@dataclass
+class FrenchGreeter(Greeter):
+    """ A customer kind of Greeter """
+
+    greeting: str = 'Bonjour'
+
+
+def setup(registry: ServiceRegistry):
+    register_dataclass(
+        registry,
+        FrenchGreeter, Greeter,
+        context=FrenchCustomer
+    )
+
+    # Grab the Datastore and add a FrenchCustomer
+    container = registry.create_container()
+    datastore: Datastore = container.get(Datastore)
+    henri = FrenchCustomer(name='henri', title='Henri')
+    datastore.customers['henri'] = henri
```

Comment:
what can we do to get rid of this config-time container? To me, the idea of injecting data into this datastore is something that is not related to setup - it's related to the fact that the datastore here is memory-only instead of persistent. Shouldn't this be done somewhere else where the datastore is created - import `FrenchCustomer` and make one and add it?

`setup(registry)` to me is about registering services and not about adding data to singletons.

Reply:
The idea here is that an add-on might want to do some post-registry stuff. It would be like a Pyramid add-on wanting an RDBMS table.

I could split that into two steps, but I'd like to show the add-on handling things itself, instead of either the app or the "site".

Reply:
My concern is that it's just not conforming to the intent of the api. It requires specific buy-in from the caller to know that the `setup()` function is going to do this, and it **only** works if the datastore was registered as a singleton. It'd be way more obvious from an api and contractual standpoint to just pass the datastore in via `setup(registry, datastore)` instead of just assuming that everything you're doing is with a singleton.

Reply:
If wired provided a way to know that the datastore was a singleton it might feel a bit better but I can't think of a way to make that work comfortably either. The intent is always that you don't care inside a container because the mutations you do are all to the local container and may be thrown away after the container is thrown away - or at least you acknowledge its out of your hands what happens at that point.

Reply:
How about showing two phases:

- `register` does what you're discussing, populate the registry, no more no less
- `post_register` hands the registry to an add-on to do whatever the hell it wants. And since the Datastore is now reachable, they can get the datastore.

Most importantly, I should emphasize that we aren't shipping an app. They write whatever handshake they want, because we are promoting custom apps.

Reply:
So if the app is programmed around the idea of a global datastore then that thing should be passed into `setup()` right? This is the kind of thing a configurator solves where the addon is one-hop-abstracted from the actual work. The addon declares it wants to do things, then the app does them. For example `config.add_customer(FrenchCustomer(name='henri', title='Henri'), name='henri')`. This signals intent without making the addon responsible for how it happens.

Reply:
At this point I need to declare defeat. It's too complicated for a "how to write your own DI system" tutorial. I wanted to show what "context" was good for and thought the Pyramid pattern of context being a DB row would be familiar.

I've gone too far in the direction of writing a configurator.

Reply:
Maybe it's time to start exploring how to write a general-purpose configurator. I think that's the idea behind dectate, I'm not familiar with too many others. Several other users have asked over the years about pulling the configurator core out of Pyramid.

For this tutorial what about just `setup(registry, datastore)` or something like that, which is simpler to grok for users even if it's not perfect?

Reply:
"For this tutorial what about just setup(registry, datastore) or something like that even if it's not perfect? Is that a deal breaker for you?"

That's fine with me. I thought you wanted that thing to also have a declarative system where the app was applying the add-on's instructions to the datastore. Just to make clear, you're saying:

- Don't do a two-phase setup, keep the current one-phase, but...
- Explicitly pass in the datastore to the add-on's setup, and thus...
- ...add-on's in this example don't need to make a container

Is that right?

Reply:
That's my proposal yes. You need to tell me if that simplification conflicts with the goals of your tutorial. It's basically making the `setup()` api customized on a per-app basis.

Reply:
Doesn't conflict at all. I want to show some good patterns but at the same time, want to show people get to make their own thing.

I'll make the change. Thanks again for taking the time, sorry about the messiness.

Reply:
Done, makes more sense this way.